### DISCUSSION TITLE
Is there a way to disable bodies dynamically at run-time?

### DISCUSSION THREAD

**Discussion by rpapallas** [google-deepmind/mujoco#439] [Asking for Help] [answered]
I am using MuJoCo with the Python API and while I load a model with multiple bodies, I would like at run-time to disable/remove some objects. When I say disabling/removing objects, I mean to make MuJoCo not simulate the dynamics of those objects at every time step. I experimented with disabling contacts by setting the `contype` and `conaffinity` of the bodies, but this has no computational benefit to MuJoCo.  Is there a better way to tell MuJoCo "I want to remove this object from the model" or "don't simulate this object at all" that will have an actual computational benefit to MuJoCo if I do so?

Thank you.

**Comment by yuvaltassa** [comment, marked as answer]
While we plan to address this use case in the future, the current answer is No, you cannot do that.

For some situations you can put an object far away and then bring it back to where you want it as a proxy for actually removing it. I realise this is a hacky solution that may not apply to your use case.

**Reply by rpapallas** [reply]
Thank you for your reply.

**Reply by rpapallas** [reply]
I realize that when I move objects away from interactions, as you suggest, versus removing the same objects from the XML model entirely, MuJoCo dynamics are different. Imagine a robot arm interacting with objects, when you remove *some* objects from XML as opposed to moving the same objects *far* away from the arm interactions, but still part of the model, gives different dynamics. That has been puzzling me for a while. I don't think I have something wrong in my code, and I am wondering if MuJoCo optimisers behind the scene are behaving differently when objects are removed from the model as opposed to moving the same objects far away?

Edit: I experiment with changing the XML model and moving 40 objects far away from an object and the robot, then execute a trajectory and observed where the object pushed by the robot landed. Then removed the 40 objects, leaving only the robot and that single object, executed the same trajectory and the object landed slightly differently than before. The only change is having those 40 objects far away, vs not having them at all.

If you also think this is not expected, I am happy to open a bug report and provide more details of how to reproduce it.

**Reply by yuvaltassa** [reply]
This is to be expected. I assume you are executing an open-loop trajectory?
When contacts are involved, any numerical difference will soon accumulate and be visible as trajectory divergence (usually around the time contacts happen). This is due to the very high Lyapunov exponents created by contacts. Since having the other objects in the scene changes the numerics of the solver, you will see this divergence. This will happen in any simulator, not just MuJoCo.
However, if you execute the same trajectory twice in the same model, you will get perfect numerical equivalence.
